### ISSUE TITLE
Can select a range with blocked dates in it

### ISSUE THREAD

**Issue by majapw** [airbnb/react-dates#23]
This should maybe be an option instead of the default. We're going to eventually need to prevent people from selecting date ranges that include blocked dates.

**Comment by agupta93** [comment]
@majapw can you please specify the exact behaviour that is expected, as in when the user tries to click on end date with block dates included in duration should the user click be disabled or is he allowed to click but the date range is not selected with an error saying blocked dates in range ?

**Comment by majapw** [comment]
I think no error message for now (we're working on a possible design for that sort of scenario). Most importantly, if there is a hover span, that should disappear once the span itself contains a blocked date (similar to what happens when you hover over a blocked date itself). 

As for the click, the click should register up through to the `DateRangePicker` component, but should be prevented from doing anything further in that component's `handleDayClick` method I think. 

My biggest problem with this when I tried to implement this previously was performance. Namely I got a solution that worked rather well for short ranges, but slowed the entire component down hella when the range was a handful of months.

**Comment by agupta93** [comment]
@majapw okay i kinda understood it will start working on this one now :) will ask if i encounter further doubts

**Comment by agupta93** [comment]
@majapw i have tried implementing this feature and as you said there were performance issues as the range increased to months and years. This majorly stems from the fact that you have to iterate though days again and again. I have tried implementing in a way to avoid such iteration and have reached to the following behaviour screenshot is attached 

![image](https://cloud.githubusercontent.com/assets/7349199/19019421/54e7db68-88a5-11e6-90d0-7c167de5a24f.png)

As it can be seen in the image that if blocked dates are not allowed the span only extends before the first blocked date and disappears on any date after the blocked that indicating that the user can only select the range without blocked dates. Tell me if this behaviour is ok so that i can proceed with it further

**Comment by gigitux** [comment]
Any update?

**Comment by oldo** [comment]
Looks great @agupta93 
This is exactly the behaviour that we are looking for. Do you have the work you've done so far on a branch/fork that I might be able to checkout?

**Comment by agupta93** [comment]
@oldo yes i might have the branch but as u can see it was done months ago, since i received no confirmation i stopped working on it. Its still a work in progress tell me when you need it and will try to complete and give you the branch :)

**Comment by oldo** [comment]
Ready when you are @agupta93 :-)
I'm about to start working on implementing this feature now so even if you are not 100% happy with what you've done please feel free to share it as it would be a great headstart.
Thanks!

**Comment by agupta93** [comment]
@oldo okay will give u the branch this weekend busy in some lately i hope u can wait :)

**Comment by oldo** [comment]
@agupta93 thank you! I look forward to having a look :-)

**Comment by agupta93** [comment]
@oldo https://github.com/agupta93/react-dates/tree/blocked-dates

**Comment by vongkim** [comment]
Are there any updates to this?

**Comment by oldo** [comment]
@agupta93 apologies for the delayed response. Thanks very much for sharing your work. In the end I resolved the problem by creating a function that checks for blocked dates between a `start` and `end` date:

```javascript
export function checkForBlockedDates(start, end, dates) {
  const dateFormat = 'YYYY-MM-DD';
  const diff = moment(end).diff(start, 'days') + 1;

  for (let i = 0; i < diff; i++) {
    const checkDate = moment(start).add(i, 'd').format(dateFormat);

    if (dates[checkDate] && dates[checkDate].blocked) {
      return true;
    }
  }

  return false;
}
```

And if this function returns `true` then I just clear selection and display a modal saying that there are blocked dates within the selected range.

**Comment by ingoclaro** [comment]
my $0.02: After the user selects the first date, disable all dates after the first blocked date following the selected date. A different background color could be used to differentiate these dates from the blocked dates.

**Comment by timhwang21** [comment]
@ingoclaro agreed, that's how we're doing it on our end. But there are some undesirable edge cases, because if there is no initial value, the initial check for blocked dates after start date is selected happens before `startDate` is passed to the change handler.

@majapw is there still desire for a PR on this? Thanks!

**Comment by laurenz-glueck** [comment]
@agupta93 would be really helpful if this feature would be included as an option into the master branch - any plans for that?

**Comment by hemanth-sp** [comment]
how add this your code in angular 2 ???????????

**Comment by ljharb** [comment]
@hemanth-sp as you asked in #1326, this is REACT dates. It only ever works in react, and will not work in any past, present, or future version of angular.

**Comment by colemerrick** [comment]
My $.02 - think there should be a clear difference between blocked functionally and blocked visually.  Perhaps 2 props - isDayBlocked vs isDayDisabled?  With isDayDisabled not preventing a range selection with a disabled day in the middle.  

Use case: bookings at businesses/schools where one needs to disable weekends/holidays but still need to select date-ranges over them.

**Comment by hemanth-sp** [comment]
i understand , consider hotel booking system, on boy wants select sat,sun
and monday but sunday hotel is closed so how to do that in this situation?

**Comment by janeschindler** [comment]
This is also a featured I'd love to see.

**Comment by heleneshaikh** [comment]
I think that this is a very important and basic feature to implement.

**Comment by skevlishvili** [comment]
This would be a great feature to have

**Comment by giverz** [comment]
As a workaround I'm comparing the startDate to the next earliest blocked date, and passing these dates to isOutsideRange. One downside is that the function doesn't get called when the user backspace-deletes the Check In field for example.

Would love to have this as a feature!

**Comment by lianju** [comment]
Any updates here? 3 years went by since pull request.

**Comment by lianju** [comment]
@majapw How do you handle this at airbnb? Seems that it’s not possible to select a range with blocked dates.

**Comment by IadityaM** [comment]
Any updates on this? Could really use this functionality.

**Comment by omeiirr** [comment]
Hi, any updates?
Is this a part of the main branch yet?
It still allows the user to select a range containing some blocked date.

@oldo Thanks for the function you described above. How can I use that? Where should it be called? A simple example would be of immense help. Thanks. 😁

**Comment by oldo** [comment]
Ufff, sorry @omeiirr, I posted over 4 years ago! Haven't used this library in a while but from memory it was just called when end date is set. Sorry I'm not much more help :-)

**Comment by ryanwoconnor** [comment]
Has this been resolved? This feels like a pretty important option to have.

**Comment by multiwebinc** [comment]
Are you looking for a PR to do this?

**Comment by ljharb** [comment]
@multiwebinc yep, that’s what the “pull request wanted” label means.

**Comment by danishfareed** [comment]
I think the end is soon for this lib, as there's only 1 sole maintainer for this lib.

**Comment by ljharb** [comment]
@danishfareed that's hostile and unhelpful, and inaccurate.
